File: sudoku_backtracking.py
import numpy as np
from sudoku_helper import is_valid_placement

def solve_sudoku_backtracking(sudoku_grid):
    def backtrack(sudoku_grid):
        for row in range(9):
            for col in range(9):
                if sudoku_grid[row, col] == 0:
                    for num in range(1, 10):
                        if is_valid_placement(sudoku_grid, row, col, num):
                            sudoku_grid[row, col] = num
                            if backtrack(sudoku_grid):
                                return True
                            sudoku_grid[row, col] = 0
                    return False
        return True
    backtrack(sudoku_grid)

def solve_sudoku_backtracking_counting_nodes(sudoku_grid):
    node_counter = [0]

    def backtrack(sudoku_grid):
        node_counter[0] += 1
        for row in range(9):
            for col in range(9):
                if sudoku_grid[row, col] == 0:
                    for num in range(1, 10):
                        if is_valid_placement(sudoku_grid, row, col, num):
                            sudoku_grid[row, col] = num
                            if backtrack(sudoku_grid):
                                return True
                            sudoku_grid[row, col] = 0
                    return False
        return True

    backtrack(sudoku_grid)
    
    return node_counter[0]

File: typescript/sudokuBacktracking.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.solveSudokuBacktracking = solveSudokuBacktracking;
const sudokuHelper_1 = require("./sudokuHelper");
function solveSudokuBacktracking(sudokuGrid) {
    function backtrack(sudokuGrid) {
        for (let row = 0; row < 9; row++) {
            for (let col = 0; col < 9; col++) {
                if (sudokuGrid[row][col] === 0) {
                    for (let num = 1; num < 10; num++) {
                        if ((0, sudokuHelper_1.isValidPlacement)(sudokuGrid, row, col, num)) {
                            sudokuGrid[row][col] = num;
                            if (backtrack(sudokuGrid))
                                return true;
                            sudokuGrid[row][col] = 0;
                        }
                    }
                    return false;
                }
            }
        }
        return true;
    }
    backtrack(sudokuGrid);
}
